esp/services/ws_workunits/ws_workunitsHelpers.cpp

```diff
@@ -2089,53 +2020,98 @@ void WsWuInfo::getWorkunitThorSlaveLog(const char *groupName, const char *ipAddr
+    getClusterThorGroupName(groupName, instanceName);
+    if (groupName.isEmpty())
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed to get Thor Group Name for %s", instanceName);
+
+    Owned<IGroup> nodeGroup = queryNamedGroupStore().lookup(groupName);
+    if (!nodeGroup || (nodeGroup->ordinality() == 0))
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Node group %s not found", groupName.str());
+
+    getWorkunitThorSlaveLog(nodeGroup, ipAddress, logDate, logDir.str(), slaveNum, buf, outFile, forDownload);
+}
+
+void WsWuInfo::readWorkunitLog(IFileIOStream* ios, MemoryBuffer& buf, const char* outFile)
+{
+    VStringBuffer startwuid("Started wuid=%s", wuid.str());
+    VStringBuffer endwuid("Finished wuid=%s", wuid.str());
 
-        VStringBuffer startwuid("Started wuid=%s", wuid.str());
-        VStringBuffer endwuid("Finished wuid=%s", wuid.str());
+    bool eof = false;
+    bool outputThisLine = false;
+    unsigned processID = 0;
+    StringBuffer line;
 
-        const char *sw = startwuid.str();
-        const char *ew = endwuid.str();
+    Owned<IFileIOStream> outIOS;
+    if (!isEmptyString(outFile))
+    {
+        CWsWuFileHelper helper(nullptr);
+        outIOS.setown(helper.createIOStreamWithFileName(outFile, IFOcreate));
+    }
 
-        Owned<IFileIOStream> outIOS;
-        if (!isEmptyString(outFile))
+    Owned<IStreamLineReader> lineReader = createLineReader(ios, true);
+    while (!eof)
+    {
+        eof = lineReader->readLine(line.clear());
+        if (strstr(line, startwuid))
         {
-            CWsWuFileHelper helper(nullptr);
-            outIOS.setown(helper.createIOStreamWithFileName(outFile, IFOcreate));
+            outputThisLine = true;
+            outputALine(line.length(), line.str(), buf, outIOS);
+            processID = readLogProcessID(line);
         }
-        while (!eof)
+        else if (outputThisLine)
         {
-            line.clear();
-            for (;;)
-            {
-                char c;
-                size32_t numRead = ios->read(1, &c);
-                if (!numRead)
-                {
-                    eof = true;
-                    break;
-                }
-                line.append(c);
-                if (c=='\n')
-                    break;
-            }
-            if (strstr(line.str(), sw))
-                include = true;
-            if (include)
-                outputALine(line.length(), line.str(), buf, outIOS);
-            if (strstr(line.str(), ew))
-                include = false;
+            //If the slave is restarted before WU is finished, we cannot find out the "Finished wuid=...".
+            //So, we should check whether the slave is restarting or not.
+            unsigned pID = readLogProcessID(line);
+            if ((pID > 0) && (pID != processID))
+                break;
+
+            outputALine(line.length(), line.str(), buf, outIOS);
         }
+
+        if (strstr(line, endwuid))
+            break;
```

Comment:
@jakesmith are you saying that, in one log file, there may be >1 wuid blocks and each of them is started by 'Started wuid=same id' and ended by 'Finished wuid=same id'.

Reply:
Yes. The existing code (in master) appears deliberately coded to handle it.

Reply:
Ok

Reply:
I do find >1 wuid blocks in one slave log using my latest code. So, we have to scan the whole file and it would take more time to create a ZAP report.

Reply:
>I do find >1 wuid blocks in one slave log using my latest code

You mean in pending changes not pushed yet?

There's no choice (just as before), re. reading whole file (as before) - you don't know whether there will be another 'Started wuid=' match or not.
If you don't there will be bits missing (most of the logging in fact typically, as most workunits have multiple graphs in the same log).
Well you *could* in theory by reading the workunit and looking at the timings and deduce how many more there were, but that would only be of marginal help in general.

Reply:
I haven't pushed the changes yet. I want to find a good time to test since it takes a long time.

I thought one WU has one Start/Finish block. I did not realize that every graph has one block.

Do you think it would save some time if I add whole log file (copyFile) without checking the Start/Finish?

Reply:
it shouldn't, I don't think that would be a good idea.
I have some planned changes that are going to separate out each workunit logging into separate logs. When that's done, no searching will be required and the whole file can be copied.

But for now with this PR, the main benefit should be the thread pool, the search for every graph was/still is required.